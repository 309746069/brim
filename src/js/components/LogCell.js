import React from "react"
import DownArrow from "../icons/chevron-bottom-md.svg"
import FieldFilter from "../models/FieldFilter"
import {ContextMenu, MenuItem} from "./ContextMenu"

export default class LogCell extends React.PureComponent {
  constructor(props) {
    super(props)
    this.state = {showMenu: false}
    this.include = this.include.bind(this)
    this.exclude = this.exclude.bind(this)
    this.toggleMenu = () => this.setState({showMenu: !this.state.showMenu})
  }

  include(_e) {
    const filter = new FieldFilter({
      name: this.props.name,
      value: this.props.value,
      operator: "="
    })

    this.props.appendToQuery(filter.toProgramFragment())
  }

  exclude(_e) {
    const filter = new FieldFilter({
      name: this.props.name,
      value: this.props.value,
      operator: "!="
    })

    this.props.appendToQuery(filter.toProgramFragment())
  }

  countBy() {
    this.props.appendToQuery(`| count() by ${this.props.name}`)
  }

  render() {
    const {name, type, value, appendQueryInclude} = this.props
    let cellClass = `log-cell ${type}`
    if (this.state.showMenu) cellClass += " active"

    let valueClass = ""

    if (name === "_path") {
      valueClass += ` ${name} ${value}-bg-color`
    }

    return (
      <div className={cellClass}>
        <span className={valueClass}>{value}</span>
        <span className="field-name">{name}</span>
        <button className="cell-options-button" onClick={this.toggleMenu}>
          <DownArrow />
        </button>
        {this.state.showMenu && (
          <ContextMenu onOutsideClick={this.toggleMenu}>
            <MenuItem onClick={this.exclude}>Filter out these values</MenuItem>
            <MenuItem onClick={() => appendQueryInclude(name, value)}>
              Only show these values
            </MenuItem>
            <MenuItem onClick={this.countBy}>Count by this field</MenuItem>
          </ContextMenu>
        )}
      </div>
    )
  }
}

export const TsCell = ({ts, highlight, onClick}) => {
  return (
    <div
      className={`ts-cell ${highlight ? "highlight" : ""}`}
      onClick={onClick}
    >
      <p>
        <span className="date">{ts.format("MMM DD, YYYY")}</span>
        <span className="time">{ts.format("HH:mm")}</span>
        <span className="seconds">{ts.format("ss.SSSS[s]")}</span>
      </p>
    </div>
  )
}
