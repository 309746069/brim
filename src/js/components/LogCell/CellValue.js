/* @flow */

import React from "react"

import type {$Field} from "../../brim"
import {format} from "../../lib/Time"

type Props = {field: $Field}

const CellValue = ({field}: Props) => {
  if (field.name === "_path")
    return (
      <span className={`${field.name} ${field.value}-bg-color `}>
        {field.display()}
      </span>
    )
  if (field.type === "time")
    return (
      <div>
        <span className="date">{format(field.toDate(), "MM/DD/YY")}</span>
        <span className="time">{format(field.toDate(), "HH:mm")}</span>
        <span className="seconds">{format(field.toDate(), "ss.SSSS")}</span>
      </div>
    )

  return <span>{field.display()}</span>
}

export default CellValue
