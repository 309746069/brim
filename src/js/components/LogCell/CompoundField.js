/* @flow */

import React from "react"
import classNames from "classnames"

import {type $Field} from "../../brim"
import Log from "../../models/Log"
import SingleField from "./SingleField"

type Props = {
  field: $Field,
  log: Log,
  menuBuilder: Function
}
export default function CompoundField({field, log, menuBuilder}: Props) {
  let compound = field.toCompound()
  let render = []

  for (let i = 0; i < compound.length; ++i) {
    let item = compound.item(i)
    if (item) {
      let menu = menuBuilder(item, log, true)
      render.push(<SingleField key={i} field={item} log={log} menu={menu} />)
    }
    if (i !== compound.length - 1) {
      render.push(<Comma />)
    }
  }

  return <Wrapper type={compound.container}>{render}</Wrapper>
}

function Comma() {
  return <Extra value="," className="separator" />
}

function Extra({value, className}) {
  return (
    <div className={classNames("compound-field-extra", className)}>{value}</div>
  )
}

function Wrapper({type, children}) {
  let [open, close] = getWrapper(type)
  return (
    <>
      <Extra value={open} />
      {children}
      <Extra value={close} />
    </>
  )
}

function getWrapper(container) {
  switch (container) {
    case "set":
      return ["{", "}"]
    case "vector":
      return ["[", "]"]
    default:
      return [null, null]
  }
}
