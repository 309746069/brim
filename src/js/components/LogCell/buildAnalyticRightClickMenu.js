/* @flow */

import {analyticDetail, seperator, whois} from "./rightClick"
import Field from "../../models/Field"
import Log from "../../models/Log"

type Args = {
  log: Log,
  field: Field
}

export default ({log, field}: Args) => {
  const menu = []

  if (["addr", "set[addr]"].includes(field.type)) {
    menu.push(whois(field))
    menu.push(seperator())
  }

  menu.push(analyticDetail(log))

  return menu
}
