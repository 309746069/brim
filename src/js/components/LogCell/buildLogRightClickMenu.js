/* @flow */

import type {Dispatch} from "../../reducers/types"
import type {Space} from "../../lib/Space"
import {
  countBy,
  detail,
  exclude,
  fromTime,
  include,
  pcaps,
  seperator,
  toTime,
  whois
} from "./rightClick"
import Field, {TimeField} from "../../models/Field"
import Log from "../../models/Log"

type Action = {
  type: "action",
  text: string,
  onClick: (Dispatch, Event) => void
}

type Seperator = {
  type: "seperator"
}

type Args = {
  field: Field,
  log: Log,
  space: Space
}

export type MenuItemData = Seperator | Action

export default (args: Args): MenuItemData[] => [
  ...fieldActions(args),
  seperator(),
  ...logActionsFunc(args)
]

const fieldActions = ({field}) => {
  const menu = []
  if (field instanceof TimeField) {
    menu.push(fromTime(field))
    menu.push(toTime(field))
  } else {
    menu.push(exclude(field))
    menu.push(include(field))
    menu.push(countBy(field))
  }

  if (["addr", "set[addr]"].includes(field.type)) {
    menu.push(seperator())
    menu.push(whois(field))
  }

  return menu
}

const logActionsFunc = ({log, space}) => {
  if (log.isPath("conn") && space.packet_support) {
    return [pcaps(log), detail(log)]
  } else {
    return [detail(log)]
  }
}
