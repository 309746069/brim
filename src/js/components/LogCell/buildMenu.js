/* @flow */

import * as searchBar from "../../actions/searchBar"
import * as packets from "../../actions/packets"
import * as logDetails from "../../actions/logDetails"
import * as timeWindow from "../../actions/timeWindow"
import type {Props} from "./LogCellActions"
import {TimeField} from "../../models/Field"

type Action = {
  type: "action",
  text: string,
  onClick: Event => void
}

type Seperator = {
  type: "seperator"
}

export type MenuItemData = Seperator | Action

export default ({dispatch, field, space, log}: Props): MenuItemData[] => {
  if (log.isPath("conn") && space.packet_support) {
    return [
      exclude(field, dispatch),
      include(field, dispatch),
      countBy(field, dispatch),
      seperator(),
      detail(log, dispatch),
      pcaps(log, dispatch)
    ]
  }

  if (field instanceof TimeField) {
    return [
      fromTime(field, dispatch),
      toTime(field, dispatch),
      seperator(),
      detail(log, dispatch)
    ]
  }

  return [
    exclude(field, dispatch),
    include(field, dispatch),
    countBy(field, dispatch),
    seperator(),
    detail(log, dispatch)
  ]
}

const exclude = (field, dispatch) => ({
  type: "action",
  text: "Filter out these values",
  onClick: (e: Event) => {
    e.stopPropagation()
    dispatch(searchBar.appendQueryExclude(field))
    dispatch(searchBar.submitSearchBar())
  }
})

const include = (field, dispatch) => ({
  type: "action",
  text: "Only show these values",
  onClick: (e: Event) => {
    e.stopPropagation()
    dispatch(searchBar.appendQueryInclude(field))
    dispatch(searchBar.submitSearchBar())
  }
})

const countBy = (field, dispatch) => ({
  type: "action",
  text: `Count by ${field.name}`,
  onClick: (e: Event) => {
    e.stopPropagation()
    dispatch(searchBar.appendQueryCountBy(field))
    dispatch(searchBar.submitSearchBar())
  }
})

const pcaps = (log, dispatch) => ({
  type: "action",
  text: "Download PCAPS",
  onClick: (_e: Event) => {
    dispatch(packets.fetchPackets(log))
  }
})

const detail = (log, dispatch) => ({
  type: "action",
  text: "View log details",
  onClick: () => {
    dispatch(logDetails.viewLogDetail(log))
  }
})

const fromTime = (field, dispatch) => ({
  type: "action",
  text: "Use as \"start\" time",
  onClick: () => {
    dispatch(timeWindow.setOuterFromTime(field.toDate()))
    dispatch(searchBar.submitSearchBar())
  }
})

const toTime = (field, dispatch) => ({
  type: "action",
  text: "Use as \"end\" time",
  onClick: () => {
    dispatch(timeWindow.setOuterToTime(field.toDate()))
    dispatch(searchBar.submitSearchBar())
  }
})

const seperator = () => ({
  type: "seperator"
})
