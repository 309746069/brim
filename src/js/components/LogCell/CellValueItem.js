/* @flow */
import React from "react"

import type {RightClickBuilder} from "../../types"
import {showContextMenu} from "../../lib/System"
import CellValue from "./CellValue"
import Field from "../../models/Field"
import Log from "../../models/Log"
import lib from "../../lib"

type Props = {
  field: Field,
  log: Log,
  valueIndex: null | number,
  menu?: RightClickBuilder
}

export default function CellValueItem({log, field, _valueIndex, menu}: Props) {
  function onContextMenu() {
    menu && showContextMenu(menu(field, log))
  }

  return (
    <div
      className="cell-value-item"
      onClick={(e) => lib.win.selectText(e.currentTarget)}
      onContextMenu={onContextMenu}
    >
      <CellValue field={field} />
    </div>
  )
}
