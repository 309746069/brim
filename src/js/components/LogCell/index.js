/* @flow */

import React, {useState} from "react"
import classNames from "classnames"

import type {RightClickBuilder} from "../../types"
import {getTooltipStyle} from "../../lib/MenuStyler"
import CellValueItem from "./CellValueItem"
import Field from "../../models/Field"
import Log from "../../models/Log"
import Tooltip from "../Tooltip"

type Props = {
  field: Field,
  log: Log,
  style?: Object,
  rightClick?: RightClickBuilder
}

export default function LogCell(props: Props) {
  let [hover, setHover] = useState(false)
  let [tooltipStyle, setTooltipStyle] = useState({})

  function handleMouseEnter(e) {
    setHover(true)
    setTooltipStyle(getTooltipStyle(e.currentTarget))
  }

  function handleMouseLeave() {
    setHover(false)
  }

  let {name, type} = props.field

  return (
    <div
      className={classNames(`log-cell ${type}`, {hover})}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      style={props.style}
    >
      <CellValueItem
        menu={props.rightClick}
        field={props.field}
        log={props.log}
        valueIndex={0}
      />
      {hover && (
        <Tooltip style={tooltipStyle}>
          <span className="field-name">{name}</span>
        </Tooltip>
      )}
    </div>
  )
}
