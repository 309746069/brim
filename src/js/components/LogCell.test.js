/* @flow */

import React from "react"
import {shallow} from "enzyme"
import LogCell from "./LogCell"
import * as mockLogs from "../test/mockLogs"

const log = mockLogs.conn()

test("shapshot", () => {
  const logCell = shallow(
    <LogCell log={log} field={log.getField("_path")} isScrolling={false} />
  )
  expect(logCell).toMatchSnapshot()
})

test("shapshot after right-clicking", () => {
  const logCell = shallow(
    <LogCell log={log} field={log.getField("_path")} isScrolling={false} />
  )
  logCell.simulate("ContextMenu", new Event("ContextMenu"))
  expect(logCell).toMatchSnapshot()
})

test("shapshot after hovering", () => {
  const logCell = shallow(
    <LogCell log={log} field={log.getField("_path")} isScrolling={false} />
  )
  logCell.simulate("MouseEnter", {
    currentTarget: {getBoundingClientRect: () => ({top: 100, left: 100})}
  })
  expect(logCell).toMatchSnapshot()
  logCell.simulate("MouseLeave")
  expect(logCell).toMatchSnapshot()
})

test("shapshot after clicking", () => {
  const logCell = shallow(
    <LogCell log={log} field={log.getField("_path")} isScrolling={false} />
  )
  logCell.simulate("Click", {})
  expect(logCell).toMatchSnapshot()
})

test("shapshot right clicking a time cell", () => {
  const logCell = shallow(
    <LogCell log={log} field={log.getField("ts")} isScrolling={false} />
  )
  logCell.simulate("ContextMenu", new Event("ContextMenu"))
  expect(logCell).toMatchSnapshot()
})
