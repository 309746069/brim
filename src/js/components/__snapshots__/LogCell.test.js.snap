// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shapshot 1`] = `
<div
  className="log-cell string"
  onClick={[Function]}
  onContextMenu={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span
    className="_path conn-bg-color "
  >
    conn
  </span>
</div>
`;

exports[`shapshot after clicking 1`] = `
<div
  className="log-cell string"
  onClick={[Function]}
  onContextMenu={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span
    className="_path conn-bg-color "
  >
    conn
  </span>
</div>
`;

exports[`shapshot after hovering 1`] = `
<div
  className="log-cell string hover"
  onClick={[Function]}
  onContextMenu={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span
    className="_path conn-bg-color "
  >
    conn
  </span>
  <button
    className="cell-options-button"
  >
    <DownArrow
      data-container-transform="translate(0 1)"
      data-icon="chevron-bottom"
      height="32"
      viewBox="0 0 32 32"
      width="32"
      xmlns="http://www.w3.org/2000/svg"
    />
  </button>
  <Tooltip
    style={Object {}}
  >
    <span
      className="field-name"
    >
      _path
    </span>
  </Tooltip>
</div>
`;

exports[`shapshot after hovering 2`] = `
<div
  className="log-cell string"
  onClick={[Function]}
  onContextMenu={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span
    className="_path conn-bg-color "
  >
    conn
  </span>
</div>
`;

exports[`shapshot after right-clicking 1`] = `
<div
  className="log-cell string active hover"
  onClick={[Function]}
  onContextMenu={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span
    className="_path conn-bg-color "
  >
    conn
  </span>
  <button
    className="cell-options-button"
  >
    <DownArrow
      data-container-transform="translate(0 1)"
      data-icon="chevron-bottom"
      height="32"
      viewBox="0 0 32 32"
      width="32"
      xmlns="http://www.w3.org/2000/svg"
    />
  </button>
  <Tooltip
    style={Object {}}
  >
    <span
      className="field-name"
    >
      _path
    </span>
  </Tooltip>
  <Connect(LogCellActions)
    field={
      Field {
        "name": "_path",
        "type": "string",
        "value": "conn",
      }
    }
    log={
      Log {
        "descriptor": Array [
          Object {
            "name": "_td",
            "type": "int",
          },
          Object {
            "name": "_path",
            "type": "string",
          },
          Object {
            "name": "ts",
            "type": "time",
          },
          Object {
            "name": "uid",
            "type": "string",
          },
          Object {
            "name": "id.orig_h",
            "type": "addr",
          },
          Object {
            "name": "id.orig_p",
            "type": "port",
          },
          Object {
            "name": "id.resp_h",
            "type": "addr",
          },
          Object {
            "name": "id.resp_p",
            "type": "port",
          },
          Object {
            "name": "proto",
            "type": "enum",
          },
          Object {
            "name": "service",
            "type": "string",
          },
          Object {
            "name": "duration",
            "type": "interval",
          },
          Object {
            "name": "orig_bytes",
            "type": "count",
          },
          Object {
            "name": "resp_bytes",
            "type": "count",
          },
          Object {
            "name": "conn_state",
            "type": "string",
          },
          Object {
            "name": "local_orig",
            "type": "bool",
          },
          Object {
            "name": "local_resp",
            "type": "bool",
          },
          Object {
            "name": "missed_bytes",
            "type": "count",
          },
          Object {
            "name": "history",
            "type": "string",
          },
          Object {
            "name": "orig_pkts",
            "type": "count",
          },
          Object {
            "name": "orig_ip_bytes",
            "type": "count",
          },
          Object {
            "name": "resp_pkts",
            "type": "count",
          },
          Object {
            "name": "resp_ip_bytes",
            "type": "count",
          },
          Object {
            "name": "tunnel_parents",
            "type": "set[string]",
          },
        ],
        "tuple": Array [
          "0",
          "conn",
          "1425612054.369843",
          "Cynwae4qh1GxM82hQ2",
          "192.168.0.50",
          "1900",
          "239.255.255.250",
          "1900",
          "udp",
          "-",
          "2.000293",
          "282",
          "0",
          "S0",
          "-",
          "-",
          "0",
          "D",
          "3",
          "366",
          "0",
          "0",
          "(empty)",
        ],
      }
    }
    onClose={[Function]}
    style={
      Object {
        "left": undefined,
        "top": undefined,
      }
    }
  />
</div>
`;

exports[`shapshot right clicking a time cell 1`] = `
<div
  className="log-cell time active hover"
  onClick={[Function]}
  onContextMenu={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <p>
    <span
      className="date"
    >
      03/05/15
    </span>
    <span
      className="time"
    >
      19:20
    </span>
    <span
      className="seconds"
    >
      54.3690
    </span>
  </p>
  <button
    className="cell-options-button"
  >
    <DownArrow
      data-container-transform="translate(0 1)"
      data-icon="chevron-bottom"
      height="32"
      viewBox="0 0 32 32"
      width="32"
      xmlns="http://www.w3.org/2000/svg"
    />
  </button>
  <Tooltip
    style={Object {}}
  >
    <span
      className="field-name"
    >
      ts
    </span>
  </Tooltip>
  <Connect(LogCellActions)
    field={
      TimeField {
        "name": "ts",
        "type": "time",
        "value": "1425612054.369843",
      }
    }
    log={
      Log {
        "descriptor": Array [
          Object {
            "name": "_td",
            "type": "int",
          },
          Object {
            "name": "_path",
            "type": "string",
          },
          Object {
            "name": "ts",
            "type": "time",
          },
          Object {
            "name": "uid",
            "type": "string",
          },
          Object {
            "name": "id.orig_h",
            "type": "addr",
          },
          Object {
            "name": "id.orig_p",
            "type": "port",
          },
          Object {
            "name": "id.resp_h",
            "type": "addr",
          },
          Object {
            "name": "id.resp_p",
            "type": "port",
          },
          Object {
            "name": "proto",
            "type": "enum",
          },
          Object {
            "name": "service",
            "type": "string",
          },
          Object {
            "name": "duration",
            "type": "interval",
          },
          Object {
            "name": "orig_bytes",
            "type": "count",
          },
          Object {
            "name": "resp_bytes",
            "type": "count",
          },
          Object {
            "name": "conn_state",
            "type": "string",
          },
          Object {
            "name": "local_orig",
            "type": "bool",
          },
          Object {
            "name": "local_resp",
            "type": "bool",
          },
          Object {
            "name": "missed_bytes",
            "type": "count",
          },
          Object {
            "name": "history",
            "type": "string",
          },
          Object {
            "name": "orig_pkts",
            "type": "count",
          },
          Object {
            "name": "orig_ip_bytes",
            "type": "count",
          },
          Object {
            "name": "resp_pkts",
            "type": "count",
          },
          Object {
            "name": "resp_ip_bytes",
            "type": "count",
          },
          Object {
            "name": "tunnel_parents",
            "type": "set[string]",
          },
        ],
        "tuple": Array [
          "0",
          "conn",
          "1425612054.369843",
          "Cynwae4qh1GxM82hQ2",
          "192.168.0.50",
          "1900",
          "239.255.255.250",
          "1900",
          "udp",
          "-",
          "2.000293",
          "282",
          "0",
          "S0",
          "-",
          "-",
          "0",
          "D",
          "3",
          "366",
          "0",
          "0",
          "(empty)",
        ],
      }
    }
    onClose={[Function]}
    style={
      Object {
        "left": undefined,
        "top": undefined,
      }
    }
  />
</div>
`;
